File: python/프로그래머스/석유시추.py
from collections import deque

def solution(land):
    answer = []
    n = len(land)
    m = len(land[0])
    check = [[-1 for _ in range(m)] for _ in range(n)]
    visited = [[False for _ in range(m)] for _ in range(n)]
    def bfs(x,y,check,idx):
        moves = [(0,1), (0,-1),(1,0),(-1,0)]
        n = len(check)
        m = len(check[0])
        q = deque()
        q.append([x , y])
        visited[x][y] = True
        count = 1 
        check[x][y] = idx
        while q:
            x, y = q.popleft()
            
            for px,py in moves:
                if 0 <= x + px < n and 0 <= y + py < m:
                    if land[x+px][y+py] == 1 and not visited[x+px][y+py]:
                        q.append([x+px, y+py])
                        visited[x+px][y+py] = True
                        check[x+px][y+py] = idx
                        count += 1
        return count
    idx = 0
    value = [0 for _ in range(n*m)]
    for x in range(n):
        for y in range(m):
            if land[x][y] == 1 and not visited[x][y]:
                temp = bfs(x,y,check,idx)
                value[idx] = temp
                idx += 1
    
    for y in range(m):
        visit = []
        for x in range(n):
            v = check[x][y]
            if v != -1:
                visit.append(v)
        temp_result = 0
        for a in set(visit):
            temp_result += value[a]
        answer.append(temp_result)

    return max(answer)

File: java/석유시추.java
import java.util.*;

class Solution {
    int[][] moves = new int[][]{{1,0},{-1,0},{0,1},{0,-1}};
    int xx;
    int yy;
    public int solution(int[][] land) {
        int answer = 0;
        xx = land.length;
        yy = land[0].length;
        boolean[][] visited = new boolean[xx][yy];
        Map<Integer, Integer> map = new HashMap<>();
        int idx = 2;
        Set<Integer> set = new HashSet<>();
        Set<Integer> make = new HashSet<>();
        for(int i=0;i<yy;i++) {
            int cnt = 0;
            for(int j=0;j<xx;j++) {
                int key = land[j][i];
                if(key>0) {
                    int v = 0;
                    if(key==1){
                        v = dfs(land,visited,j, i, 1 ,idx);
                        map.put(idx, v);
                        make.add(idx++);
                        cnt += v;
                    } else if(!make.contains(key)) {
                        set.add(key);
                    }
                }
            }
            
            for(int k: set) 
                cnt += map.get(k);    
            set.clear();
            make.clear();
            answer = Math.max(cnt, answer);
        }

        return answer;
    } 
    
    int dfs(int[][] land, boolean[][] visited, int x, int y, int count, int idx) {
        visited[x][y] = true;
        land[x][y] = idx;
        for(int[]mv: moves) {
            int nx = x + mv[0];
            int ny = y + mv[1];
            
            if(0<=nx&&nx<xx&&0<=ny&&ny<yy){
                if(!visited[nx][ny] && land[nx][ny]==1) {
                    count = dfs(land, visited, nx, ny, count+1, idx);
                }
            }
        }
        
        return count;
    }
}
